### ISSUE TITLE
Derivation `outputs` order

### ISSUE THREAD

**Issue by CobaltCause** [cargo2nix/cargo2nix#279]
In the readme and the examples, this pattern can be observed:

```nix
hello-world = (rustPkgs.workspace.hello-world {}).bin;
#                                                ^^^^ this part
```

If the `.bin` is intended to be used by default, is there any reason why `outputs` is defined like this?

https://github.com/cargo2nix/cargo2nix/blob/c149357cc3d17f2849c73eb7a09d07a307cdcfe8/overlay/mkcrate.nix#L319

In particular, putting `"bin"` first would make that the default so the `.bin` stuff could be omitted. From [the manual](https://nixos.org/manual/nix/stable/expressions/derivations.html):

> The first element of `outputs` determines the *default output*.

I found it useful to swap the order [here, for example](https://or.computer.surgery/charles/cargo-shim/-/blob/4357ead02d7a63593bcd8f9b8cb413beca599d5c/flake.nix#L39-53), but maybe I'm just doing something wrong.

**Comment by psionic-k** [comment]
Takes 4 lines.  I had to make the tera output use the `out` attribute explicitly, but other than that, it looks like a naive swap.  I'll get it into the unstable branch.  Right now I'm focused on fixing features for target-specific dependencies / features.  That will be the 0.12.0 release pretty soon.

**Comment by CobaltCause** [comment]
Nice!

I'm wondering if, since the `bin` output only includes `$bin/bin/{the,built,binaries}` while `out` contains `$out/{bin,lib,nix-support}`, if `"bin"` should only come first for top-level builds, since it seems like the stuff in `out` contains the `rlib`s which I'm guessing are used for dependency-building stuff.

**Comment by psionic-k** [comment]
AFAIK "top-level builds" are derivations not referenced downstream in dependent derivations.  We don't know when we build one if it's top level or not.  Re-ordering the outputs depending on the situation would create two unique derivations, but if we only use one of the outputs from both, it's virtually the same as having two outputs.  The difference is choosing outputs versus choosing derivations.

The oddball for our output is that we include not just rlibs but linking info that needs replayed downstream.  As things are, a crate derivation doesn't know its linking info up front.  It gets it by looking at the output of the upstream dependencies.  This is what a lot of behavior in mkcrate.nix actually is, recording and replaying linking information.

**Comment by psionic-k** [comment]
Closing this.  Please use the `unstable` branch to include changes discussed in this issue.

Fix will ship in 0.12.0 release as well.
